src/passes/LLVMNontrappingFPToIntLowering.cpp

```diff
@@ -0,0 +1,150 @@
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+#include <limits>
+
+// By default LLVM emits nontrapping float-to-int instructions to implement its
+// fptoui/fptosi conversion instructions. This pass replaces these instructions
+// with code sequences which also implement LLVM's fptoui/fptosi, but which are
+// not semantically equivalent in wasm. This is because out-of-range inputs to
+// these instructions produce poison values. So we need only ensure that there
+// is no trap, but need not ensure any particular result.
+
+namespace wasm {
+struct LLVMNonTrappingFPToIntLowering
+  : public WalkerPass<PostWalker<LLVMNonTrappingFPToIntLowering>> {
+
+  UnaryOp getReplacementOp(UnaryOp op) {
+    switch (op) {
+      case TruncSatSFloat32ToInt32:
+        return UnaryOp::TruncSFloat32ToInt32;
+      case TruncSatUFloat32ToInt32:
+        return UnaryOp::TruncUFloat32ToInt32;
+      case TruncSatSFloat64ToInt32:
+        return UnaryOp::TruncSFloat64ToInt32;
+      case TruncSatUFloat64ToInt32:
+        return UnaryOp::TruncUFloat64ToInt32;
+      case TruncSatSFloat32ToInt64:
+        return UnaryOp::TruncSFloat32ToInt64;
+      case TruncSatUFloat32ToInt64:
+        return UnaryOp::TruncUFloat32ToInt64;
+      case TruncSatSFloat64ToInt64:
+        return UnaryOp::TruncSFloat64ToInt64;
+      case TruncSatUFloat64ToInt64:
+        return UnaryOp::TruncUFloat64ToInt64;
+      default:
+        WASM_UNREACHABLE("Unexpected opcode");
+    }
+  }
+
+  template<typename From, typename To> void ReplaceSigned(Unary* curr) {
+    BinaryOp ltOp;
+    UnaryOp absOp;
+    switch (curr->op) {
+      case TruncSatSFloat32ToInt32:
+      case TruncSatSFloat32ToInt64:
+        ltOp = BinaryOp::LtFloat32;
+        absOp = UnaryOp::AbsFloat32;
+        break;
+      case TruncSatSFloat64ToInt32:
+      case TruncSatSFloat64ToInt64:
+        ltOp = BinaryOp::LtFloat64;
+        absOp = UnaryOp::AbsFloat64;
+        break;
+      default:
+        WASM_UNREACHABLE("Unexpected opcode");
+    }
+
+    Builder builder(*getModule());
+    replaceCurrent(builder.makeIf(
+      builder.makeBinary(
+        ltOp,
+        builder.makeUnary(absOp, curr->value),
+        builder.makeConst(static_cast<From>(std::numeric_limits<To>::max()))),
+      builder.makeUnary(getReplacementOp(curr->op), curr->value),
```

Comment:
This results in invalid IR because I'm reusing `curr->value` here and you can't have the same node duplicated in the tree. (Oops, this isn't SSA).
What's the recommended thing to do here? Just stash the incoming value in a local? (Create a new one, and run a cleanup pass afterward in case the input is already a get_local like in the test)?

Reply:
Yes, add a local and reuse the value that way.

No need to add a special cleanup. In Emscripten we run these lowerings before the main optimization pipeline, so it would handle that.

Reply:
Got it, thanks.